### ISSUE TITLE
[Feature]: Enable feedback from service "fetch()" to config_entry

### ISSUE THREAD

**Issue by dala318** [mampfes/hacs_waste_collection_schedule#2820]
### I propose a feature for:

The integration in general

### Describe your wanted feature

In my ongoing service addition #2789 you can create connection to API in two ways.

1. Re-use the API-key from phone app (then you get the collections from addresses you have registered in app).
    Works without problem, but has its limitations to bind the same addresses in integration and phone app
2. Register a new API-key and add an address to it via some extra API calls. 
   This sort of works if registrering address via GUI (you get a few errors "feedback from registration" on the way but doing the correct selections from the proposals given you get a unique API-Key stored with the wanted address)
   This on the other hand don't work if you setup the service via `configuration.yaml` (any exception fails the setup)

This request is about the (2.) alternative above.

If the call to `Source.fetch()` would include have freedom to directly affect the result (that is finally stored to the `hass.data` via `config_entry`) without needing to go via error exceptions the flow would be smoother.

I see some alternatives (have not really tested them):

- Add a new set of exceptions from `Source` that are always caught "nicely" and directly set specific value in `config_entry` and GUI
- Add an optional callback method to `fetch()` which can modify `config_entry` to be stored

**Comment by 5ila5** [comment]
> * Add a new set of exceptions from `Source` that are always caught "nicely" and directly set specific value in `config_entry` and GUI
> 
> * Add an optional callback method to `fetch()` which can modify `config_entry` to be stored

I don't think this is a good approach, I'd rather add some kind of mechanism to interact with a configuration wizard before really configuring the Source. So you would create a method that sorts everything out beforehand and returns all necessary arguments for the source, which is then configured once.

The fetch is called from yaml and GUI configurations, and adding some additional exceptions to handle to the config_flow adds unnecessary complexity. 

I hadn't had time yet to really think about how to "advertise" the wizard to the config flow (I have thought about decorators but haven't looked into it if this is really good idea.)

**Comment by markvp** [comment]
Thanks for the detailed write-up, and for the context from 5ila5 as well.

The approach of having `fetch()` write back to `config_entry` would introduce meaningful coupling between the data-fetching layer and the HA integration layer — the core library is designed to be HA-independent, so this would cut against that grain.

5ila5's suggestion of a pre-configuration wizard step is the preferred direction here. The `wizard/` directory already has examples of this pattern (e.g. `abfall_io.py`, `jumomind_de.py`) where cascading API lookups are resolved interactively before the source is configured. Extending that mechanism to cover the "register a new API key, then store it" use case would keep the architecture clean and work for both the GUI and `configuration.yaml` setups.

That said, the wizard infrastructure isn't trivial to extend for this particular use case (persisting a generated key back into the config entry after a registration API call), so there's no quick fix here. Leaving this open as a tracked enhancement — contributions welcome if anyone wants to take a crack at the wizard approach.
